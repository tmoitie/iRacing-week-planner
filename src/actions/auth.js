import {
  getAuth, signOut as fbSignOut, signInWithEmailAndPassword, createUserWithEmailAndPassword,
  onAuthStateChanged, sendPasswordResetEmail, GoogleAuthProvider, signInWithPopup,
} from 'firebase/auth';
import { debouncedDispatcherSaveSettings, getSettingsFromFirebase, saveSettingsToFirebase } from './settings';

export const LOADING_AUTH = 'AUTH/LOADING_SIGN_IN';
export const ERROR_AUTH = 'AUTH/ERROR_AUTH';
export const ERROR_RESET = 'AUTH/ERROR_RESET';
export const ERROR_ACKNOWLEDGE = 'AUTH/ERROR_ACKNOWLEDGE';
export const SIGNED_IN = 'AUTH/SIGNED_IN';
export const SIGNED_OUT = 'AUTH/SIGNED_OUT';
export const RESET_SENT = 'AUTH/RESET_SENT';
export const LOADING_RESET = 'AUTH/LOADING_RESET';

export function signOut() {
  return async (dispatch, getState) => {
    dispatch({ type: LOADING_AUTH });
    if (getState().settings.firebaseSynced === false) {
      debouncedDispatcherSaveSettings.cancel();
      await dispatch(saveSettingsToFirebase());
    }
    const auth = getAuth(getState().auth.firebaseApp);
    await fbSignOut(auth);

    dispatch({ type: SIGNED_OUT });
  };
}

export function signIn(email, password) {
  return async (dispatch, getState) => {
    dispatch({ type: LOADING_AUTH });
    try {
      const auth = getAuth(getState().auth.firebaseApp);
      await signInWithEmailAndPassword(auth, email, password);
      return {};
    } catch (error) {
      dispatch({ type: ERROR_AUTH, error });
      return { type: ERROR_AUTH };
    }
  };
}

export function signInWithGoogle() {
  return async (dispatch, getState) => {
    dispatch({ type: LOADING_AUTH });
    try {
      const auth = getAuth(getState().auth.firebaseApp);

      const provider = new GoogleAuthProvider();
      provider.addScope('profile');
      provider.addScope('email');

      await signInWithPopup(auth, provider);
      return {};
    } catch (error) {
      dispatch({ type: ERROR_AUTH, error });
      return { type: ERROR_AUTH };
    }
  };
}

export function createAccount(email, password) {
  return async (dispatch, getState) => {
    dispatch({ type: LOADING_AUTH });
    try {
      const auth = getAuth(getState().auth.firebaseApp);
      await createUserWithEmailAndPassword(auth, email, password);
      return {};
    } catch (error) {
      dispatch({ type: ERROR_AUTH, error });
      return { type: ERROR_AUTH };
    }
  };
}

export function acknowledgeAuthError() {
  return { type: ERROR_ACKNOWLEDGE };
}

export function signedIn(user) {
  return async (dispatch) => {
    await dispatch({ type: SIGNED_IN, user });

    if (user) {
      await dispatch(getSettingsFromFirebase());
    }
  };
}

export function startListener() {
  return async (dispatch, getState) => {
    const auth = getAuth(getState().auth.firebaseApp);
    const { currentUser } = auth;
    dispatch(signedIn(currentUser));

    onAuthStateChanged(auth, (user) => {
      dispatch(signedIn(user));
    });
  };
}

export function forgottenPassword(email) {
  return async (dispatch, getState) => {
    dispatch({ type: LOADING_RESET });
    try {
      const auth = getAuth(getState().auth.firebaseApp);
      await sendPasswordResetEmail(auth, email);
      dispatch({ type: RESET_SENT });
      return { type: RESET_SENT };
    } catch (error) {
      dispatch({ type: ERROR_RESET, error });
      return { type: ERROR_RESET };
    }
  };
}
