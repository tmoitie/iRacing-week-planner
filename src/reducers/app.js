// @flow

import moment, { duration } from 'moment';
import { CHANGE_MODAL, UPDATE_DAYS } from '../actions/app';
import { SIGNED_IN } from '../actions/auth';
import { seasonStart, weekSeasonStart, seasonEnd } from '../config';

const getWeek = (date) => Math.ceil(duration(moment(date).add({ second: 1 }).diff(weekSeasonStart)).asWeeks());

type AppState = {
  date: moment$Moment,
  daysSinceSeasonStart: number,
  week: number,
  currentModal: string,
};

type AppReducerArgs = {
  type?: string,
  days?: number,
  modalName?: string,
};

export default function app(
  initState: ?AppState,
  { type, days, modalName }: AppReducerArgs,
): AppState {
  let state = initState;

  if (state === undefined) {
    let currentDate = moment(new Date()).utc().startOf('day');

    if (currentDate.isBefore(seasonStart)) {
      currentDate = moment(seasonStart).utc().startOf('day');
    }

    if (currentDate.isAfter(seasonEnd)) {
      currentDate = moment(seasonEnd).utc().startOf('day').subtract({ days: 1 });
    }

    const currentDays = currentDate.diff(seasonStart, 'days');

    state = {
      date: currentDate,
      daysSinceSeasonStart: currentDays,
      week: getWeek(currentDate),
      currentModal: null,
    };
  }

  if (type === UPDATE_DAYS) {
    const date = moment(seasonStart).add(days, 'days');
    return {
      ...state,
      date,
      daysSinceSeasonStart: days,
      week: getWeek(date),
    };
  }

  if (type === CHANGE_MODAL) {
    return {
      ...state,
      currentModal: modalName,
    };
  }

  if (type === SIGNED_IN) {
    return {
      ...state,
      currentModal: null,
    };
  }

  return state;
}
