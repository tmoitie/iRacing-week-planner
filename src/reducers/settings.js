// @flow

import uniq from 'lodash.uniq';
import { actionTypes as localStorageActionTypes } from 'redux-localstorage';
import { SIGNED_OUT } from '../actions/auth';
import {
  FIREBASE_SYNCED,
  LOAD_SETTINGS_FROM_FIREBASE,
  RESET_FILTERS,
  RESET_SETTINGS,
  UPDATE_FILTERS,
  UPDATE_SETTING,
} from '../actions/settings';

import availableColumns from '../data/availableColumns';
import { tracks, cars } from '../data';

export type SortOptions = {
  key: string,
  order: string,
};

export type FilterOptions = {
  type: Array<string>,
  licence: Array<string>,
  official: Array<boolean>,
  fixed: boolean[],
  ownedCars: boolean,
  ownedTracks: boolean,
  favouriteSeries: boolean,
  favouriteCarsOnly: boolean,
  favouriteTracksOnly: boolean,
};

export const defaultFilters: FilterOptions = {
  type: ['Sports Car', 'Formula Car', 'Oval', 'Dirt', 'RX'],
  licence: ['R', 'D', 'C', 'B', 'A', 'P'],
  official: [false, true],
  fixed: [false, true],
  ownedCars: false,
  ownedTracks: false,
  favouriteSeries: false,
  favouriteTracksOnly: false,
  favouriteCarsOnly: false,
};

export type SettingOptions = {
  filters: FilterOptions,
  ownedCars: Array<number>,
  ownedTracks: Array<number>,
  favouriteSeries: Array<number>,
  favouriteCars: Array<number>,
  favouriteTracks: Array<number>,
  sort: SortOptions,
  columns: Array<number>,
  firebaseSynced: boolean,
};

export const defaultSettings: SettingOptions = {
  filters: defaultFilters,
  ownedCars: cars.filter((car) => car.freeWithSubscription === true).map((car) => car.sku),
  ownedTracks: tracks.filter((track) => track.default).map((track) => track.pkgid),
  favouriteSeries: [],
  favouriteCars: [],
  favouriteTracks: [],
  sort: { key: 'licence', order: 'asc' },
  columns: availableColumns.filter((column) => column.default === true).map((column) => column.id),
  firebaseSynced: false,
};

function synchroniseLegacyChanges(state) {
  const newState = { ...state };

  newState.ownedCars = uniq([
    ...newState.ownedCars,
    ...defaultSettings.ownedCars,
  ]);

  newState.ownedTracks = uniq([
    ...newState.ownedTracks,
    ...defaultSettings.ownedTracks,
  ]);

  const hasRoadFilter = newState.filters.type.includes('Road');

  if (hasRoadFilter) {
    newState.filters.type = [...newState.filters.type, 'Sports Car', 'Formula Car'];
    newState.filters.type = newState.filters.type.filter((t) => t !== 'Road');
  }

  return newState;
}

export default function settings(initState: SettingOptions, { type, payload }): SettingOptions {
  let state = initState;
  if (initState === undefined) {
    state = defaultSettings;
  }

  if (type === localStorageActionTypes.INIT) {
    if (!payload) {
      return state;
    }

    const newState = synchroniseLegacyChanges(payload.settings);

    return { ...state, ...newState };
  }

  if (type === UPDATE_FILTERS) {
    return { ...state, filters: payload.filters, firebaseSynced: false };
  }

  if (type === RESET_FILTERS) {
    return { ...state, filters: defaultFilters, firebaseSynced: false };
  }

  if (type === RESET_SETTINGS || type === SIGNED_OUT) {
    return { ...defaultSettings, firebaseSynced: false };
  }

  if (type === UPDATE_SETTING) {
    return { ...state, [payload.key]: payload.value, firebaseSynced: false };
  }

  if (type === LOAD_SETTINGS_FROM_FIREBASE) {
    return synchroniseLegacyChanges({
      ...state,
      ...defaultSettings,
      ...payload,
      firebaseSynced: true,
    });
  }

  if (type === FIREBASE_SYNCED) {
    return { ...state, firebaseSynced: true };
  }

  return state;
}
