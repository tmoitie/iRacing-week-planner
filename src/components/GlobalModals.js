// @flow

import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { shallowEqual, useDispatch, useSelector } from 'react-redux';
import { changeModal } from '../actions/app';
import { cars, tracks } from '../data';
import { defaultSettings } from '../reducers/settings';
import { updateSetting } from '../actions/settings';
import AboutModal from './modal/AboutModal';
import ContentModal from './modal/ContentModal';
import FavouriteSeriesModal from './modal/FavouriteSeriesModal';
import LoginModal from './modal/LoginModal';
import OptionsModal from './modal/OptionsModal';
import PurchaseGuideModal from './modal/PurchaseGuideModal';

const currentModalSelector = (state) => state.app.currentModal;
const settingsSelector = (state) => state.settings;
const sortedActiveCars = [...cars.filter((c) => !c.name.startsWith('['))].sort((a, b) => a.name.localeCompare(b.name));
const sortedLegacyCars = [...cars.filter((c) => c.name.startsWith('['))].sort((a, b) => a.name.localeCompare(b.name));
const sortedCars = [...sortedActiveCars, ...sortedLegacyCars];

const sortedActiveTracks = [
  ...tracks.filter((t) => !t.name.startsWith('[')),
].sort((a, b) => a.name.localeCompare(b.name));
const sortedLegacyTracks = [
  ...tracks.filter((t) => t.name.startsWith('[')),
].sort((a, b) => a.name.localeCompare(b.name));
const sortedTracks = [...sortedActiveTracks, ...sortedLegacyTracks];

export default function GlobalModals() {
  const { t } = useTranslation();
  const currentModal = useSelector(currentModalSelector);
  const {
    favouriteSeries, ownedTracks, favouriteTracks, ownedCars, favouriteCars, columns,
  } = useSelector(settingsSelector, shallowEqual);
  const dispatch = useDispatch();

  const closeModal = (e) => {
    if (e) {
      e.preventDefault();
    }
    dispatch(changeModal(null));
  };

  const getSettingUpdater = (key) => (value) => {
    dispatch(updateSetting(key, value));
  };

  return (
    <>
      <FavouriteSeriesModal
        isOpen={currentModal === 'favourite-series'}
        onClose={closeModal}
        favouriteSeries={favouriteSeries}
        save={getSettingUpdater('favouriteSeries')}
      />
      <ContentModal
        id="content-modal-tracks"
        isOpen={currentModal === 'my-tracks'}
        onClose={closeModal}
        title={t('Set my tracks')}
        ownedContent={ownedTracks}
        content={sortedTracks}
        idField="pkgid"
        defaultContent={[...defaultSettings.ownedTracks]}
        typeFilter={{
          key: 'primaryType',
          oval: 'oval',
          road: 'road',
          dirtRoad: 'dirt_road',
          dirtOval: 'dirt_oval',
        }}
        save={getSettingUpdater('ownedTracks')}
        favourites={favouriteTracks}
        saveFavourites={getSettingUpdater('favouriteTracks')}
      />
      <ContentModal
        id="content-modal-cars"
        isOpen={currentModal === 'my-cars'}
        onClose={closeModal}
        title={t('Set my cars')}
        ownedContent={ownedCars}
        content={sortedCars}
        idField="sku"
        defaultContent={[...defaultSettings.ownedCars]}
        typeFilter={{
          key: 'categories',
          oval: ['oval'],
          sportsCar: ['sports_car'],
          formulaCar: ['formula_car'],
          dirtRoad: ['dirt_road'],
          dirtOval: ['dirt_oval'],
        }}
        save={getSettingUpdater('ownedCars')}
        favourites={favouriteCars}
        saveFavourites={getSettingUpdater('favouriteCars')}
      />
      <OptionsModal
        isOpen={currentModal === 'options'}
        onClose={closeModal}
        columnIds={columns}
        saveOptions={(key, value) => dispatch(updateSetting(key, value))}
      />
      <AboutModal
        isOpen={currentModal === 'about'}
        onClose={closeModal}
      />
      <PurchaseGuideModal
        isOpen={currentModal === 'purchase-guide'}
        onClose={closeModal}
        ownedTracks={ownedTracks}
        favouriteSeries={favouriteSeries}
      />
      <LoginModal
        isOpen={currentModal === 'login'}
        onClose={closeModal}
      />
    </>
  );
}
