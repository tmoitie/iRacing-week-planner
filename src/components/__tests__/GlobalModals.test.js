import { describe, test } from '@jest/globals';
import React from 'react';
import { render, screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import GlobalModals from '../GlobalModals';
import { CHANGE_MODAL } from '../../actions/app';
import { UPDATE_SETTING } from '../../actions/settings';

describe('components/GlobalModals', () => {
  const mockStore = configureMockStore([thunk]);
  const defaultStore = {
    settings: {
      favouriteSeries: [],
      ownedTracks: [],
      favouriteTracks: [],
      ownedCars: [],
      favouriteCars: [],
      columns: [],
    },
    app: {
      currentModal: null,
    },
    auth: {
      user: null,
      firebaseApp: {},
    },
  };

  test('renders all closed', async () => {
    const store = mockStore(defaultStore);
    const { container } = render(<Provider store={store}><GlobalModals /></Provider>);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders open options, closing and saving', async () => {
    const user = userEvent.setup();
    let store = mockStore({ ...defaultStore, app: { currentModal: 'options' } });
    const { baseElement, rerender } = render(<Provider store={store}><GlobalModals /></Provider>);

    expect(baseElement).toMatchSnapshot();

    await user.click(within(screen.getByTestId('optionsModal')).getByLabelText('ID'));
    expect(store.getActions()).toContainEqual({ type: UPDATE_SETTING, payload: { key: 'columns', value: ['id'] } });

    await user.click(within(screen.getByTestId('optionsModal')).getByLabelText('Close'));
    expect(store.getActions()).toContainEqual({ type: CHANGE_MODAL, modalName: null });

    store = mockStore({ ...defaultStore, app: { currentModal: 'favourite-series' } });
    rerender(<Provider store={store}><GlobalModals /></Provider>);

    await user.click(within(screen.getByTestId('favouriteSeriesModal')).getByLabelText('13th Week iRacing Figure GR8'));

    expect(store.getActions()).toContainEqual({
      type: UPDATE_SETTING,
      payload: { key: 'favouriteSeries', value: [245] },
    });

    await user.click(within(screen.getByTestId('favouriteSeriesModal')).getByLabelText('Close'));
    expect(store.getActions()).toContainEqual({ type: CHANGE_MODAL, modalName: null });
  });
});
