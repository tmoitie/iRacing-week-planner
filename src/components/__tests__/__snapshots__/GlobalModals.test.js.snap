// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/GlobalModals renders all closed 1`] = `null`;

exports[`components/GlobalModals renders open options, closing and saving 1`] = `
<body>
  <div />
  <div
    id="optionsModal"
  >
    <div
      class="ReactModal__Overlay"
      style="position: fixed; display: block; inset: 0; background-color: rgba(255, 255, 255, 0.75);"
    >
      <div
        class="Modal__Bootstrap modal-dialog modal-lg"
        data-testid="optionsModal"
      >
        <div
          class="modal-content"
        >
          <div
            class="modal-header"
          >
            <button
              aria-label="Close"
              class="close"
              type="button"
            >
              <span
                class="glyphicon glyphicon-remove"
              />
            </button>
            <h4
              class="modal-title"
            >
              Options
            </h4>
          </div>
          <div
            class="modal-body"
            style="max-height: 55vh; overflow-y: auto;"
          >
            <div
              class="container-fluid"
            >
              <h5>
                Columns
              </h5>
              <div
                class="row"
              >
                <div
                  class="col-xs-3 col-md-2"
                >
                  <div
                    class="checkbox"
                    data-testid="options-columns-id"
                  >
                    <label
                      for="options-columns-id"
                    >
                      <input
                        id="options-columns-id"
                        type="checkbox"
                      />
                      ID
                    </label>
                  </div>
                </div>
                <div
                  class="col-xs-3 col-md-2"
                >
                  <div
                    class="checkbox"
                    data-testid="options-columns-class"
                  >
                    <label
                      for="options-columns-class"
                    >
                      <input
                        id="options-columns-class"
                        type="checkbox"
                      />
                      Class
                    </label>
                  </div>
                </div>
                <div
                  class="col-xs-3 col-md-2"
                >
                  <div
                    class="checkbox"
                    data-testid="options-columns-licence"
                  >
                    <label
                      for="options-columns-licence"
                    >
                      <input
                        id="options-columns-licence"
                        type="checkbox"
                      />
                      Licence
                    </label>
                  </div>
                </div>
                <div
                  class="col-xs-3 col-md-2"
                >
                  <div
                    class="checkbox"
                    data-testid="options-columns-type"
                  >
                    <label
                      for="options-columns-type"
                    >
                      <input
                        id="options-columns-type"
                        type="checkbox"
                      />
                      Type
                    </label>
                  </div>
                </div>
                <div
                  class="col-xs-3 col-md-2"
                >
                  <div
                    class="checkbox disabled"
                    data-testid="options-columns-series"
                  >
                    <label
                      for="options-columns-series"
                    >
                      <input
                        disabled=""
                        id="options-columns-series"
                        type="checkbox"
                      />
                      Series
                    </label>
                  </div>
                </div>
                <div
                  class="col-xs-3 col-md-2"
                >
                  <div
                    class="checkbox"
                    data-testid="options-columns-track"
                  >
                    <label
                      for="options-columns-track"
                    >
                      <input
                        id="options-columns-track"
                        type="checkbox"
                      />
                      Track
                    </label>
                  </div>
                </div>
                <div
                  class="col-xs-3 col-md-2"
                >
                  <div
                    class="checkbox"
                    data-testid="options-columns-car"
                  >
                    <label
                      for="options-columns-car"
                    >
                      <input
                        id="options-columns-car"
                        type="checkbox"
                      />
                      Car
                    </label>
                  </div>
                </div>
                <div
                  class="col-xs-3 col-md-2"
                >
                  <div
                    class="checkbox"
                    data-testid="options-columns-start"
                  >
                    <label
                      for="options-columns-start"
                    >
                      <input
                        id="options-columns-start"
                        type="checkbox"
                      />
                      Start
                    </label>
                  </div>
                </div>
                <div
                  class="col-xs-3 col-md-2"
                >
                  <div
                    class="checkbox"
                    data-testid="options-columns-end"
                  >
                    <label
                      for="options-columns-end"
                    >
                      <input
                        id="options-columns-end"
                        type="checkbox"
                      />
                      End
                    </label>
                  </div>
                </div>
                <div
                  class="col-xs-3 col-md-2"
                >
                  <div
                    class="checkbox"
                    data-testid="options-columns-racelength"
                  >
                    <label
                      for="options-columns-racelength"
                    >
                      <input
                        id="options-columns-racelength"
                        type="checkbox"
                      />
                      Length
                    </label>
                  </div>
                </div>
                <div
                  class="col-xs-3 col-md-2"
                >
                  <div
                    class="checkbox"
                    data-testid="options-columns-official"
                  >
                    <label
                      for="options-columns-official"
                    >
                      <input
                        id="options-columns-official"
                        type="checkbox"
                      />
                      Official
                    </label>
                  </div>
                </div>
                <div
                  class="col-xs-3 col-md-2"
                >
                  <div
                    class="checkbox"
                    data-testid="options-columns-fixed"
                  >
                    <label
                      for="options-columns-fixed"
                    >
                      <input
                        id="options-columns-fixed"
                        type="checkbox"
                      />
                      Fixed
                    </label>
                  </div>
                </div>
                <div
                  class="col-xs-3 col-md-2"
                >
                  <div
                    class="checkbox"
                    data-testid="options-columns-raceTimes"
                  >
                    <label
                      for="options-columns-raceTimes"
                    >
                      <input
                        id="options-columns-raceTimes"
                        type="checkbox"
                      />
                      Race times
                    </label>
                  </div>
                </div>
                <div
                  class="col-xs-3 col-md-2"
                >
                  <div
                    class="checkbox"
                    data-testid="options-columns-nextRace"
                  >
                    <label
                      for="options-columns-nextRace"
                    >
                      <input
                        id="options-columns-nextRace"
                        type="checkbox"
                      />
                      Next race
                    </label>
                  </div>
                </div>
                <div
                  class="col-xs-3 col-md-2"
                >
                  <div
                    class="checkbox"
                    data-testid="options-columns-seriesEnd"
                  >
                    <label
                      for="options-columns-seriesEnd"
                    >
                      <input
                        id="options-columns-seriesEnd"
                        type="checkbox"
                      />
                      Season end
                    </label>
                  </div>
                </div>
                <div
                  class="col-xs-3 col-md-2"
                >
                  <div
                    class="checkbox"
                    data-testid="options-columns-seriesLink"
                  >
                    <label
                      for="options-columns-seriesLink"
                    >
                      <input
                        id="options-columns-seriesLink"
                        type="checkbox"
                      />
                      Link
                    </label>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="modal-footer"
          >
            <button
              class="btn btn-primary"
              type="button"
            >
              Close
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
